src/Compilers/CSharp/Portable/Parser/LanguageParser.cs

```diff
@@ -6242,15 +6242,31 @@ private ScanTypeFlags ScanType(ParseTypeMode mode, out SyntaxToken lastTokenOfTy
                 }
             }
 
-            if (this.CurrentToken.Kind == SyntaxKind.IdentifierToken)
+            // Handle :: as well for error case of an alias used without a preceding identifier.
+            if (this.CurrentToken.Kind is SyntaxKind.IdentifierToken or SyntaxKind.ColonColonToken)
             {
-                result = this.ScanNamedTypePart(out lastTokenOfType);
-                if (result == ScanTypeFlags.NotType)
+                bool isAlias;
+                if (this.CurrentToken.Kind is SyntaxKind.ColonColonToken)
                 {
-                    return ScanTypeFlags.NotType;
+                    result = ScanTypeFlags.NonGenericTypeOrExpression;
+                    isAlias = true;
+
+                    // We set this to null to appease the flow checker.  It will always be the case that this will be
+                    // set to an appropriate value inside the `for` loop below.  We'll consume the :: there and then
+                    // call ScanNamedTypePart which will always set this to a valid value.
+                    lastTokenOfType = null;
                 }
+                else
+                {
+                    result = this.ScanNamedTypePart(out lastTokenOfType);
+                    if (result == ScanTypeFlags.NotType)
+                    {
+                        return ScanTypeFlags.NotType;
+                    }
 
-                bool isAlias = this.CurrentToken.Kind == SyntaxKind.ColonColonToken;
+                    Debug.Assert(result is ScanTypeFlags.GenericTypeOrExpression or ScanTypeFlags.GenericTypeOrMethod or ScanTypeFlags.NonGenericTypeOrExpression);
```

Comment:
>ScanTypeFlags.GenericTypeOrExpression or ScanTypeFlags.GenericTypeOrMethod [](start = 43, length = 74)

Do we ever hit these cases? It doesn't feel right to set isAlias to true for them even if the current token is ```::```. It feels wrong to return ScanTypeFlags.AliasQualifiedName for ```A<B>::C```. When we parse this as a local type, we do not create AliasQualifiedName.

Reply:
@AlekseyTs does 072ffbe6340 fit more what you were thinking about here?

Reply:
> does 072ffbe6340 fit more what you were thinking about here?

I think my concern is addressed now. Thanks!

---
In reply to: [573199798](https://github.com/dotnet/roslyn/pull/51028#discussion_r573199798) [](ancestors = 573199798)